CARDS-1737: When computing the surveys that need to be filled out for a visit based on the configured frequency, look at previous visit from the same clinic rather than with the same questionnaire set

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/CARDS-1737).

This PR changes the required forms that need to be filled out for a visit to be linked to the _clinic_, rather than just the questionnaire.

**To test:** Checkout the test branch at `CARDS-1737-test` before continuing, as it contains a change to `mock_torch.js` to allow it to send test data for different clinics.
** Making sure nothing broke: **
1. Under the OSGi config (for me: `http://localhost:8080/system/console/configMgr`), scroll to the bottom and create a new `PROMs import` configuration, with the following:
- Name: test
- Endpoint URL: `http://localhost:8011/`
- Clinic-name: `6012-HC-Congenital Cardiac`
- Everything else default. Save.
2. Startup nodejs mock Torch server from `Utilities/Development` with the command `nodejs mock_torch.js --appointment-time-hours-from-now=48`
3. Open up `http://localhost:8080/Subjects.importTorch.json?config=mock` in your browser. You should see "Data import started". At this point there's an error in the logs because the Authentication URL is broken, but it should otherwise work.
4. Running this import multiple times should still just result in 9 new forms
5. On the sidebar, navigating to PMCC - ACHD and checking Upcoming appointments should result in just 3 of them no matter how many times you ran the import.

![image](https://user-images.githubusercontent.com/4656440/168902953-4a0b47da-2a51-4972-95e9-e736c41a80ee.png)

**Testing the clinic-specific changes:**
6. Go to Administration->Clinics->Create a new Clinic with ID `TestMapping` and name `Test clinic mapping` and Surveys `Mood`. Save.
7. Stop the nodejs mock Torch server, and rerun it specifying your new clinic, a.k.a.: `nodejs mock_torch.js --appointment-time-hours-from-now=48 --appointment-location=TestMapping`
8. Under the OSGi config (for me: `http://localhost:8080/system/console/configMgr`), edit the latest configuration so the Clinic-name is set to `TestMapping`.
9. Open up `http://localhost:8080/Subjects.importTorch.json?config=mock` in your browser. You should see "Data import started".
10. Now there should be 15 forms (Mood has less forms)
11. Under the new Test sidebar entry, 3 new upcoming appointments should be filled out. Make sure the URLs under the Survey completion differ from the ones for PMC - ACHD
![image](https://user-images.githubusercontent.com/4656440/168903854-85894d3b-78d8-4f10-a4ee-7745a2aa8fda.png)

Anything else that looked at the old `surveys` field in `Visit Information` should still be functional, even though the field itself is now gone from imports.

OrdiNeu committed:
CARDS-1737: When computing the surveys that need to be filled out for a
visit based on the configured frequency, look at previous visit from the
same clinic rather than with the same questionnaire set

OrdiNeu committed:
CARDS-1737: Remove clinic display name functionality, change from looking at
Surveys to looking at Clinics

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/QuestionnaireSet.jsx:
```diff
@@ -226,7 +226,7 @@ function QuestionnaireSet(props) {
           setSubjectData(json);
           setVisitInformation(json[visitInformationFormTitle]?.[0] || {});
-          setId(Object.values(json[visitInformationFormTitle]?.[0]).find(o => o?.question?.text == "Surveys")?.value);
+          setId(Object.values(json[visitInformationFormTitle]?.[0]).find(o => o?.question?.text == "Clinic")?.value);
```
The line is storing the clinic id as the questionnaire set id, so when trying to display surveys I assume nothing will show up because this is not the ID of a questionnaire set.

Since I see that the `surveys` question hasn't been removed, we can keep populating the `id` from the `surveys` field of _Visit Information_ (except get it by `@name` instead of `text`):

```
setId(Object.values(json[visitInformationFormTitle]?.[0]).find(o => o?.question?.["@name"] == "surveys")?.value);
```
If we go ahead with the plan to remove `surveys` from Visit information which is now a bit redundant, then we must get the survey id from the clinic configuration instead:
```
let clinicId = Object.values(json[visitInformationFormTitle]?.[0]).find(o => o?.question?.["@name"] == "clinic")?.value;

fetch(`/Proms/ClinicMapping/${clinicId}.json`)
 [...]
 .then(json => setId(json["survey"])
```

OrdiNeu committed:
CARDS-1737: Apply suggestions by @marta-

OrdiNeu committed:
CARDS-1737: Remove surveys field from PatientLocalStorage

OrdiNeu committed:
[wip] Re-add missing things

OrdiNeu committed:
CARDS-1737: Fix local storage of clinic mapping node

OrdiNeu committed:
CARDS-1737: Fixed a bug where 64 visits would be imported instead of 9

OrdiNeu committed:
CARDS-1737: Fix Proms/Visit views

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -382,15 +384,15 @@ private void pruneQuestionnaireSetByVisit(final Node visit, final VisitInformati
                     visitDate = (Calendar) this.formUtils
                         .getValue(this.formUtils.getAnswer(form, visitInformation.getVisitDateQuestion()));
                     visitQuestionnaireSet = (String) this.formUtils
```
`visitQuestionnaireSet` needs to be renamed to `visitClinic` or `clinicForVisit`.

marta- commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Questionnaires/Visit information.xml:
```diff
@@ -38,11 +38,11 @@
+			<value>Clinic</value>
 			<type>String</type>
 		</property>
 		<property>
```
Don't make the clinic question hidden. Remove the `displayMode` property so it defaults to `input`.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/ClinicLabelProcessor.java:
```diff
@@ -0,0 +1,110 @@
+        return null;
+    }
+
+    /***
+     * Get the String of the value from a possibly multi-valued answer.
+     * In most cases, this will be the second value of a multi-valued answer.
+     *
+     * @param prop Property to extract the value from
+     * @return The value in the VALUE_POS position
+     */
+    private String getValueFromArray(final Property prop)
```
When can the answer be multi-valued?

OrdiNeu replied:
Previous implementations involved treating the `clinic` answer like how `MultipleChoice` does it -- with the second value being the real value and the first being a display value. While that's not used now, I figured it would be future-safe.

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -204,7 +204,8 @@ void storeAppointment(final JsonObject appointment, final Resource patient)
             final Resource visitInfo = getOrCreateForm(visit, "/Questionnaires/Visit information");
-            updateVisitInformationForm(visitInfo, appointment, surveyInfo.getSurveyID(), surveyInfo.getDisplayName());
+            updateVisitInformationForm(visitInfo, appointment, surveyInfo.getDisplayName(),
+                surveyInfo.getClinicPath());
```
What used to be `surveyInfo` is turning more into a `clinicInfo`. Renaming both the `SurveyInfo` class and its instances would make the code easier to follow.

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -540,11 +541,11 @@ void updatePatientInformationForm(final Resource form, final JsonObject info)
-    void updateVisitInformationForm(final Resource form, final JsonObject info, final String surveyID,
-        final String locationName)
+    void updateVisitInformationForm(final Resource form, final JsonObject info, final String locationName,
+        final String clinicPath)
```
I see we're storing the full `@path` of the clinic mapping node instead of just the node `@name` as was mentioned in the specs, which IMO is much better :+1: . I'll update the specs accordingly.

marta- reviewed (commented):
Looks good, please see some renaming suggestions below.

OrdiNeu committed:
CARDS-1737: Apply code review suggestions by @marta-

marta- commented:
@sashaandjic alternative testing instructions without `mock_torch`:
* Create several clinics using the Clinics admin screen, some with the `Cardio` surveys, some with thr `Mood` surveys
* Manually create Patient information and Visit information forms
  * in Visit information, for the `Clinic` question, enter the `@path` of the clinic (e.g. `/Proms/ClinicMapping/853703519` for `6012-HC-Congenital Cardiac`; loading `/Proms/ClinicMapping.deep.json` in a separate tab of the browser can help with easily copy-pasting the `@path` for individual clinics)
* Follow the rest of the instructions above to check that surveys are grouped on the dashboard by clinic even if multiple clinics use the same questionnaire set (Cardio or Mood)

OrdiNeu committed:
CARDS-1737: Remove the multi-valued handling code

OrdiNeu committed:
CARDS-1737: Remove the references to multivalued fields from
ClinicLabelProcessor

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/labels/ClinicLabelProcessor.java:
```diff
@@ -0,0 +1,86 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.uhndata.cards.dataentry.internal.serialize.labels;
```
This label processor is very specific to PROMs, so it should not be in the core `data-entry` module.

marta- replied:
Addressed in 6d88f7c by switching the clinic question dataType to resource and using the generic resource label processor instead of this one.

marta- reviewed (changes requested):
Retracting my previous approval because the location of a file needs to be changed.

sashaandjic commented:
The URLs under the Survey completion differ for each clinic.
**Question**: After creation of the surveys, in Visit's filed `Clinic` the previously entered `@path` is converted to `clinic name`?

marta- committed:
CARDS-1737: When computing the surveys that need to be filled out for a visit based on the configured frequency, look at previous visit from the same clinic rather than with the same questionnaire set
Display the Clinic question as a dropdown with available clinics instead of a plain input

marta- commented:
> This label processor is very specific to PROMs, so it should not be in the core `data-entry` module.

DRAFT PR #1044 implements the `Visit information/clinic` question as a resource question (PR #1043) and removes this label processor. @OrdiNeu @sashaandjic please review/test the clinic question in particular in that PR, as well as `ResourceQuestion in general in PR #1043.

marta- committed:
CARDS-1737: When computing the surveys that need to be filled out for a visit based on the configured frequency, look at previous visit from the same clinic rather than with the same questionnaire set
Bug fix: update the dashboard widget query to use the right answer type for the clinic answer

sashaandjic commented:
Unable to login without token.
In the video below, I am first successfully signing in with the token, then after signing out I attempt to sing in without token

https://user-images.githubusercontent.com/23243244/172517097-63603738-bf6e-40db-8ac7-f245951ebd4c.mp4

.

sashaandjic commented:
Not sure if this is related to this PR, but I am not getting surveys created for any newly added clinic.

sashaandjic commented:
I am unable to login with or without token to fill out surveys created for my newly added `Cardio` clinic.

OrdiNeu committed:
CARDS-1737: Fix a crash on patient login with the new Visit Information
Form

OrdiNeu commented:
@sashaandjic 
> Unable to login without token. In the video below, I am first successfully signing in with the token, then after signing out I attempt to sing in without token

Just fixed in [946ea3e](https://github.com/data-team-uhn/cards/pull/1034/commits/946ea3efb222e1df3fa42b96c63f0c7fb3f5595b) :heavy_check_mark:

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/ValidateCredentialsServlet.java:
```diff
@@ -386,7 +386,7 @@ private boolean isVisitUpcoming(final Session session, Node visitInformationForm
         Node visitQuestionnaire = getVisitInformationQuestionnaire(session);
         Map<String, Predicate<Property>> tests = Map.of(
             // Verify surveys set
```
```suggestion
            // Verify clinic set
```

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/VisitChangeListener.java:
```diff
@@ -381,16 +382,16 @@ private void pruneQuestionnaireSetByVisit(final Node visit, final VisitInformati
 
         // Ignore forms for different clinics, or forms without a clinic
-        if (visitQuestionnaireSet == null || visitInformation.getQuestionnaireSet() != visitQuestionnaireSet) {
+        if (visitClinic == null || !visitInformation.getClinicPath().equals(visitClinic)) {
```
Nice catch...

Should we check that visitClinic is blank instead of null?

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -685,17 +686,27 @@ public String getDisplayName()
         {
             return this.displayName;
         }
+
+        public void setSurveyName(final String id)
+        {
+            this.surveyID = id;
+        }
+
+        public String getSurveyID()
+        {
+            return this.surveyID;
+        }
```
They should both be either `[gs]etSurveyName` or `[gs]etSurveyId` for consistency.

OrdiNeu committed:
CARDS-1737: Apply code review suggestions by @marta-

OrdiNeu committed:
CARDS-1737: Apply code review suggestions by @marta-

marta- reviewed (approved):
The code looks right. Pre-approval awaiting confirmation by testing.

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/PatientLocalStorage.java:
```diff
@@ -685,17 +686,27 @@ public String getDisplayName()
             return this.displayName;
         }
+
+        public void setSurveyID(final String id)
+        {
+            this.surveyID = id;
+        }
+
+        public String getSurveyID()
+        {
+            return this.surveyID;
+        }
```
Can we rename `surveyID` to `surveyId` for consistency with... everything else?

sashaandjic reviewed (approved):
tested

OrdiNeu committed:
CARDS-1737: Rename surveyID to surveyId
